packages/rules/src/mustSatisfyPeerDependencies.ts

```diff
@@ -0,0 +1,287 @@
+ * - `^15.2.1`
+ *
+ * More info: https://docs.npmjs.com/about-semantic-versioning
+ */
+export const MATCH_MAJOR_VERSION_RANGE = /^(\^?\d+|\^?\d+\.x|\^?\d+\.x\.x|\^\d+\.\d+|\^\d+\.\d+\.x|\^\d+\.\d+\.\d+)$/;
+
+// does not currently accept `<`, `>`, `=`, or `-` (e.g. `>= 1.5.2 < 2` / `1.0.0 - 1.2.0`)
+// TODO: accept minor pins `~4.2.1`
+export const RANGE_REGEX = /^(\*|x|\^?\d+(\.x|\.x\.x|\.\d+|\.\d+\.x|\.\d+\.\d+)?( \|\| \^?\d+(\.x|\.x\.x|\.\d+|\.\d+\.x|\.\d+\.\d+)?)*)$/;
+
+function checkSatisfyPeerDependencies(context: Context, opts: Options) {
+  const { skipUnparseableRanges } = opts;
+  const graphService = new PackageDependencyGraphService();
+  const packageNode = graphService.buildDependencyGraph(path.resolve(context.getPackageJsonPath()), 1);
+  const packageDependencies = packageNode.packageJson.dependencies || {};
+  const packagePeerDependencies = packageNode.packageJson.peerDependencies || {};
+  const packageJsonPath = packageNode.paths.packageJsonPath;
+  const packageName = packageNode.packageJson.name || packageJsonPath;
+
+  // check that no peer dependencies are also declared as regular dependencies
+  for (const [peerDependencyName, peerDependencyRange] of Object.entries(packagePeerDependencies)) {
+    const dependencyRange = packageDependencies[peerDependencyName];
+    if (dependencyRange != null) {
+      context.addError({
+        file: packageJsonPath,
+        message: `[0] Package ${packageName} has overloaded ${peerDependencyName} dependencies.`,
+        longMessage: `Peer dependency '${peerDependencyRange}' and regular dependency '${dependencyRange}'.`,
+      });
+    }
+  }
+
+  interface IPeerDependencyRequirement {
+    node: IPackageDependencyGraphNode;
+    range: string;
+  }
+  // map of all inherited peer dependency requirements
+  const allRequiredPeerDependencies: { [peerDependencyName: string]: IPeerDependencyRequirement[] } = {};
+
+  // for each of this package's dependencies, add the dependency's peer requirements into `allRequiredPeerDependencies`
+  for (const [, dependencyNode] of packageNode.dependencies) {
+    const requiredPeerDependencies = dependencyNode.packageJson.peerDependencies;
+    if (requiredPeerDependencies == null) {
+      continue;
+    }
+    for (const [peerDependencyName, range] of Object.entries(requiredPeerDependencies)) {
+      if (!RANGE_REGEX.test(range)) {
+        const message = `Unable to parse ${dependencyNode.packageJson.name}'s ${peerDependencyName} peer dependency range '${range}'.`;
+        if (skipUnparseableRanges) {
+          context.addWarning({
+            file: dependencyNode.paths.packageJsonPath,
+            message,
+          });
+          continue;
+        }
+        throw new Error(message);
+      }
+      if (allRequiredPeerDependencies[peerDependencyName] == null) {
+        allRequiredPeerDependencies[peerDependencyName] = [];
+      }
+      allRequiredPeerDependencies[peerDependencyName].push({ node: dependencyNode, range });
```

Comment:
if we're assuming that these ranges are parseable semvers because of the regex above, it feels like we should put the parsing logic here (closer to where we are certain its parseable due to regex check directly above), and then leave `doesASatisfyB` strictly as comparison logic

Reply:
I added a typecast

Reply:
for the record, there are currently 130 tests for `doesASatisfyB`